src/workqueue/workqueue.c

```diff
@@ -0,0 +1,849 @@
+/*
+ * Copyright (c) 2008-2017 Apple Inc. All rights reserved.
+ *
+ * @APPLE_APACHE_LICENSE_HEADER_START@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * @APPLE_APACHE_LICENSE_HEADER_END@
+ */
+
+#include "internal.h"
+
+#if HAVE_INTERNAL_PTHREAD_WORKQUEUE
+
+/*
+ * pthread_workqueue schedules and manages the execution
+ * of the work enqueued on libdispatch's queues.
+ *
+ * It provides a priority-ordered set of "overcommit" queues
+ * and a priority-ordered set of "normal" queues each served
+ * by a distinct pool of worker threads. The key distinction between
+ * overcommit and normal queues is that when work is added to an
+ * overcommit queue, pthread_workqueue will always ensure that a worker
+ * thread is available to execute it (even if this entails adding
+ * an additional worker thread to the overcommit pool). In contrast,
+ * the normal pool is managed to keep the number of runnable worker
+ * threads between a low and high watermark as long as sufficient
+ * work is available.
+ *
+ * Dynamic management of the normal thread pool is implemented using either
+ * (a) via low-frequency user-level approximation of the number of runnable
+ * normal worker threads via reading the /proc file system or (b) by
+ * a kernel extension that hooks the process change handler to accurately
+ * track the number of runnable normal worker threads. In either case,
+ * the management subsystem will respond to being below the target low
+ * watermark by waking an existing suspended worker thread or spawning an
+ * additional one if no suspended workers are available and the total
+ * number of normal workers is below the allowed maximum number of workers.
+ * The management subsystem will respond to being above the target high
+ * watermark by enqueuing a high-priority workitem that causes the worker
+ * thread that dequeues it to suspend itself.
+ */
+
+
+/////////////////////////////////
```

Comment:
we never use `////` like that, the rule is `/* */` for things outside of functions and `//` comment when inline inside a function

Here what you mean is:

```
#pragma mark - ...
```

which is an Xcode thing but is easy to wire in vim ;)